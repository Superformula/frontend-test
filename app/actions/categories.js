function loadCategoriesAction() {
    return { type: 'LOAD_CATEGORIES' };
}

function loadedCategories(categories) {
    return { type: 'LOADED_CATEGORIES', categories };
}

function loadCategoriesError(error) {
    return { type: 'LOAD_CATEGORIES_ERROR', error };
}

export function loadCategories() {
    return dispatch => {
        dispatch(loadCategoriesAction());
        return fetch('/api/categories')
            .then(async res => dispatch(loadedCategories((await res.json()).categories
                .filter(c => c.parent_aliases.indexOf('restaurants') !== -1))))
            .catch(err => dispatch(loadCategoriesError(err)));
    };
}

const defaultNotice = Object.freeze({
    categories: [],
    error: undefined,
    loading: false
});

export function reducer(state = defaultNotice, action) {
    switch (action.type) {
        case 'LOAD_CATEGORIES':
            return { ...state, loading: true, error: undefined };
        case 'LOADED_CATEGORIES':
            return { ...state, categories: action.categories, loading: false };
        case 'LOAD_CATEGORIES_ERROR':
            return { ...state, error: action.error, loading: false };
    }
    return state;
}
