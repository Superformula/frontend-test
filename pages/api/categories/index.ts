import { NextApiRequest, NextApiResponse } from 'next';
import { CategoryService } from '~/services';

const handler = async (_req: NextApiRequest, res: NextApiResponse) => {
  const data = await CategoryService.getAll();
  res.send(data);
};

export default handler;
