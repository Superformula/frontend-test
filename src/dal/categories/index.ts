import { gql, QueryResult } from "@apollo/client";
import { GetCategoriesQuery, Category } from "../../generated/graphql";

export { useGetCategoriesQuery } from "../../generated/graphql";
export type { Category } from "../../generated/graphql";

// We need to find all the categories for restaurants,
// but the categories queries don't allow this type of
// search, so what we do is we search all the business
// that match with the search them "restaurant" and we
// extract the categories out of them.
export const GET_CATEGORIES = gql`
  query GetCategories($location: String!) {
    search(term: "restaurants", location: $location) {
      business {
        categories {
          alias
          title
        }
      }
    }
  }
`;

export function getUniqueCategories(
  data?: GetCategoriesQuery
): Record<string, Category> {
  const businessList = data?.search?.business || [];
  const totalCategories: Record<string, Category> = {};
  for (const business of businessList) {
    const categories = business?.categories;
    if (!categories) {
      continue;
    }

    for (const cat of categories) {
      if (!cat) {
        continue;
      }

      totalCategories[cat.alias!] = cat;
    }
  }

  return totalCategories;
}
