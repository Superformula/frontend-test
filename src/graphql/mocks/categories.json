{
  "categories": {
    "__typename": "Categories",
    "category": [
      {
        "__typename": "Category",
        "alias": "afghani",
        "title": "Afghan",
        "parent_categories": [
          {
            "__typename": "Category",
            "alias": "restaurants",
            "title": "Restaurants"
          }
        ]
      },
      {
        "__typename": "Category",
        "alias": "african",
        "title": "African",
        "parent_categories": [
          {
            "__typename": "Category",
            "alias": "restaurants",
            "title": "Restaurants"
          }
        ]
      },
      {
        "__typename": "Category",
        "alias": "arabian",
        "title": "Arabian",
        "parent_categories": [
          {
            "__typename": "Category",
            "alias": "restaurants",
            "title": "Restaurants"
          }
        ]
      },
      {
        "__typename": "Category",
        "alias": "argentine",
        "title": "Argentine",
        "parent_categories": [
          {
            "__typename": "Category",
            "alias": "restaurants",
            "title": "Restaurants"
          }
        ]
      },
      {
        "__typename": "Category",
        "alias": "armenian",
        "title": "Armenian",
        "parent_categories": [
          {
            "__typename": "Category",
            "alias": "restaurants",
            "title": "Restaurants"
          }
        ]
      }
    ]
  }
}
