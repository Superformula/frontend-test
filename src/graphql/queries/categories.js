import { gql } from '@apollo/client';

export const categoriesQuery = gql`
  query GetCategories {
    categories(country: "US") {
      category {
        alias
        title
        parent_categories {
          alias
          title
        }
      }
    }
  }
`;
