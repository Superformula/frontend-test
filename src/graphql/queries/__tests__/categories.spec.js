import { categoriesQuery } from '../categories';

describe('categoriesQuery', () => {
  it('should build correct restaurant query', () => {
    expect(categoriesQuery).toMatchSnapshot();
  });
});
