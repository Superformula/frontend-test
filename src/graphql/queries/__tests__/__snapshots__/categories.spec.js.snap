// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`categoriesQuery should build correct restaurant query 1`] = `
Object {
  "definitions": Array [
    Object {
      "directives": Array [],
      "kind": "OperationDefinition",
      "name": Object {
        "kind": "Name",
        "value": "GetCategories",
      },
      "operation": "query",
      "selectionSet": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "alias": undefined,
            "arguments": Array [
              Object {
                "kind": "Argument",
                "name": Object {
                  "kind": "Name",
                  "value": "country",
                },
                "value": Object {
                  "block": false,
                  "kind": "StringValue",
                  "value": "US",
                },
              },
            ],
            "directives": Array [],
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "categories",
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "category",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "alias",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "title",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "parent_categories",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "alias",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "title",
                              },
                              "selectionSet": undefined,
                            },
                          ],
                        },
                      },
                    ],
                  },
                },
              ],
            },
          },
        ],
      },
      "variableDefinitions": Array [],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 190,
    "start": 0,
  },
}
`;
